keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1603 @@
+```
+
+#### Filter rules
+
+The `iptables` mode uses the `filter` table for three kinds of rules:
+
+##### Dropping or rejecting packets for services with no endpoints
+
+As with service dispatch, this is easily handled with a verdict map:
+
+```
+map no_endpoint_services {
+  type ipv4_addr . inet_proto . inet_service : verdict
+  elements = {
+    192.168.99.22 . tcp . 80 : drop,
+    172.30.0.46 . tcp . 80 : goto reject_chain,
+    1.2.3.4 . tcp . 80 : drop
+  }
+}
+
+chain filter {
+  ...
+  ip daddr . ip protocol . th dport vmap @no_endpoint_services
+  ...
+}
+
+# helper chain needed because "reject" is not a "verdict" and so can't
+# be used directly in a verdict map
+chain reject_chain {
+  reject
+}
+```
+
+##### Dropping traffic rejected by `LoadBalancerSourceRanges`
+
+The implementation of LoadBalancer source ranges will be similar to
+the ipset-based implementation in the `ipvs` kube proxy: we use one
+set to recognize "traffic that is subject to source ranges", and then
+another to recognize "traffic that is _accepted_ by its service's
+source ranges". Traffic which matches the first set but not the second
+gets dropped:
+
+```
+set firewall {
+  comment "destinations that are subject to LoadBalancerSourceRanges";
+  type ipv4_addr . inet_proto . inet_service
+}
+set firewall_allow {
+  comment "destination+sources that are allowed by LoadBalancerSourceRanges";
+  type ipv4_addr . inet_proto . inet_service . ipv4_addr
+}
+
+chain filter {
+  ...
+  ip daddr . ip protocol . th dport @firewall jump firewall_check
+  ...
+}
+
+chain firewall_check {
+  ip daddr . ip protocol . th dport . ip saddr @firewall_allow return
+  drop
+}
+```
+
+Where, eg, adding a Service with LoadBalancer IP `10.1.2.3`, port
+`80`, and source ranges `["192.168.0.3/32", "192.168.1.0/24"]` would
+result in:
+
+```
+add element ip kube_proxy firewall { 10.1.2.3 . tcp . 80 }
+add element ip kube_proxy firewall { 10.1.2.3 . tcp . 80 }
+add element ip kube_proxy firewall_allow { 10.1.2.3 . tcp . 80 . 192.168.0.3/32 }
+add element ip kube_proxy firewall_allow { 10.1.2.3 . tcp . 80 . 192.168.1.0/24 }
+```
+
+##### Forcing traffic on `HealthCheckNodePorts` to be accepted
+
+The `iptables` mode adds rules to ensure that traffic to NodePort
+services' health check ports is allowed through the firewall. eg:
+
+```
+-A KUBE-NODEPORTS -m comment --comment "ns2/svc2:p80 health check node port" -m tcp -p tcp --dport 30000 -j ACCEPT
+```
+
+(There are also rules to accept any traffic that has already been
+tagged by conntrack.)
+
+This cannot be done reliably in nftables; the `accept` and `drop`
+rules work differently than they do in iptables, and so if there is a
+firewall that would drop traffic to that port, then there is no
+guaranteed way to "sneak behind its back" like you can in iptables; we
+would need to actually properly configure _that firewall_ to accept
+the packets.
+
+However, these sorts of rules are somewhat legacy anyway; they work
```

Comment:
i think we have e2es for these in 
```
                                   
2708   ginkgo.It("should fail health check node port if there are only terminating endpoints [Feature:ProxyTerminatingEndpoints]", func(ctx context.Context) {
2709     // windows kube-proxy does not support this feature yet                                                                                                                                                               
2710     e2eskipper.SkipIfNodeOSDistroIs("windows")                                  
2711                                                                                 
2712     nodes, err := e2enode.GetBoundedReadySchedulableNodes(ctx, cs, 2)           
2713     framework.ExpectNoError(err)                                                
2714     nodeCounts := len(nodes.Items)                                              
2715     if nodeCounts < 2 {                                                         
2716       e2eskipper.Skipf("The test requires at least two ready nodes on %s, but found %v", framework.TestContext.Provider, nodeCounts)
2717     }                                                                           
2718     node0 := nodes.Items[0]                         
```

does that mean that those tests will break in clusters where this *used to* work bc of the god-power of iptables?

Reply:
That is testing that the HealthCheckNodePort answers "healthy"/"not-healthy" correctly.

The thing that I was talking about is that with the iptables proxy, someone could configure their iptables-based firewall to drop all packets, but HealthCheckNodePorts would still work, because iptables kube-proxy adds an explicit `ACCEPT` rule to override any firewall `DROP` rules.

With nftables, that won't work; if the firewall table says to drop all packets, you can't override that from the `kube-proxy` table. The administrator would have to reconfigure their firewall to say "drop all packets except for packets in the NodePort range" instead. (Like they already have to do if they are configuring a cloud-network-level firewall rather than a node-level iptables firewall.)

Reply:
yup so my q was 
- if admin is dropping those packets
- and upgrade from iptables to the -> nft proxy 

... will this (or other e2es) start breaking ?

Reply:
Ah, that's not really a question about the e2e tests; the important question is "if the admin is dropping those packets, and they switch from iptables to nftables, _will HealthCheckNodePorts start breaking_?" And the currently-planned answer to that is "yes", as explained in the section you're commenting on. The plan is for kube-proxy to stop trying to work around incorrect firewall rules, and instead require that the firewall rules actually be correct.

For this and other reasons, as discussed in the KEP, the nftables mode would not be a drop-in replacement for iptables. It may require manual configuration changes.